### PR TITLE
chore(build): add changelog plugin

### PR DESCRIPTION
**Pull request by wmontwe**
This pull request introduces a new Gradle plugin for automated, component-local changelog management based on Conventional Commits. It also adds documentation and enforces commit message conventions with a GitHub Workflow. 

The main focus is on enabling maintainers to generate and finalize changelogs per component, using commit history and standardized formatting.

### CODE REVIEW

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitClient.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+        pb.redirectErrorStream(true)
+        val proc = pb.start()
+        val out = proc.inputStream.bufferedReader().readLines()
+        val exit = proc.waitFor()
+        if (exit != 0) {
+            logWarn("[changelog] Command failed ($exit): ${cmd.joinToString(" ")}")
+            emptyList()
+        } else {
+            out
+        }
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitClient.kt]:
```suggestion
        proc.inputStream.bufferedReader().use { reader ->
            val out = reader.readLines()
            val exit = proc.waitFor()
            if (exit != 0) {
                logWarn("[changelog] Command failed ($exit): ${cmd.joinToString(" ")}")
                emptyList()
            } else {
                out
            }
        }
```

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/ChangelogPlugin.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+            val root = rootProject.projectDir
+            val versionDir = FileHelper.locateNearestVersionDir(start, root)
+
+            if (versionDir == null) {
+                logger.warn(
+                    "[changelog] No version.properties found between ${start.path} and ${root.path}. " +
+                        "CHANGELOG.md will not be created or updated.",
+                )
+            } else {
+                logger.info("[changelog] Using folder at: ${versionDir.path} for component changelog")
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/ChangelogPlugin.kt]: Is the intention to display this message everytime we run this plugin and the `versionDir` is present? If so, we need to use `lifecycle` instead, otherwise it will only display when we change the log level to info.

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/internal/GitVersionReader.kt [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+        return try {
+            val process = ProcessBuilder(command)
+                .redirectErrorStream(true)
+                .start()
+            val output = process.inputStream.bufferedReader().readLines()
+            if (process.waitFor() == 0) {
+                output.joinToString("\n")
+            } else {
+                ""
+            }
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/internal/GitVersionReader.kt]:
```suggestion
            process.inputStream.bufferedReader().use { buffer ->
                val output = buffer.readLines()
                if (process.waitFor() == 0) {
                    output.joinToString("\n")
                } else {
                    ""
                }
            }
```

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/internal/Version.kt [hunk 4]
```diff
@@ -4,15 +4,12 @@ internal data class Version(
     val major: Int,
     val minor: Int,
     val patch: Int,
-    val snapshot: Boolean,
 ) {
     fun toStringValue(): String = buildString {
         append(major).append('.').append(minor).append('.').append(patch)
-        if (snapshot) append("-SNAPSHOT")
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/internal/Version.kt:11]: Why are we removing the `-SNAPSHOT`?

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/internal/VersionManager.kt [hunk 5]
```diff
@@ -43,21 +41,18 @@ internal class VersionManager(
...
             val candidate = File(dir, FILE_NAME)
             if (candidate.exists()) {
-                // Log the located version.properties file for visibility
-                println("[versioning] Using version file: ${candidate.path}")
                 return candidate
             }
             if (dir == repoRoot) break
             dir = dir.parentFile
         }
-        println(
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/internal/VersionManager.kt:58]: I would keep this log, but using logger.warn instead.

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/CreateReleaseTagTask.kt [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    private fun runGit(repoRoot: File, vararg args: String) {
+        val command = listOf("git", "-C", repoRoot.absolutePath) + args
+        val process = ProcessBuilder(command)
+            .redirectErrorStream(true)
+            .start()
+        val output = process.inputStream.bufferedReader().readText()
+        val exitCode = process.waitFor()
+        check(exitCode == 0) {
+            "Command failed ($exitCode): ${command.joinToString(" ")}\n$output"
+        }
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/CreateReleaseTagTask.kt]:
```suggestion
        process.inputStream.bufferedReader().use { buffer ->
            val output = buffer.readText()
            val exitCode = process.waitFor()
            check(exitCode == 0) {
                "Command failed ($exitCode): ${command.joinToString(" ")}\n$output"
            }
        }
```

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/PrintVersionTask.kt [hunk 7]
```diff
@@ -27,8 +28,10 @@ abstract class PrintVersionTask : DefaultTask() {
             root = root,
         )
         val version = versionManager.get()
+        val versionFile = versionManager.sourceFile()
+            ?: error("No version.properties file found to print the project version.")
 
-        println(version.toStringValue())
+        println(GitVersionReader().read(root, versionFile, version))
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/PrintVersionTask.kt]:
```suggestion
        logger.lifecycle(GitVersionReader().read(root, versionFile, version))
```

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParser.kt [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+    private fun mapTypeToSectionType(type: String): SectionType? = when (type) {
+        "feat" -> SectionType.Features
+        "fix" -> SectionType.BugFixes
+        "docs" -> SectionType.Documentation
+        "style" -> SectionType.Styles
+        "refactor", "perf" -> SectionType.Refactoring
+        "test" -> SectionType.Tests
+        "chore", "build", "ci", "deps" -> SectionType.Chores
+        "revert" -> SectionType.Reverts
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParser.kt:54]: Should we allow/include `uplift` commits here?

File: build-plugin/plugin/src/test/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParserTest.kt [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+    fun `given invalid commit when parsing then returns null`() {
+        // Arrange
+        val message = "not a conventional commit"
+
+        // Act
+        val result = parser.parse(message)
+
+        // Assert
+        assertThat(result).isNull()
+    }
```
Comment (rafaeltonholo) [inline build-plugin/plugin/src/test/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParserTest.kt:126]: NIT: I added a few more test coverage here to ensure we cover more edge cases

```suggestion
    }

    // region Multiline commits

    @Test
    fun `given multiline commit following conventional format when parsing then parses header and ignores body`() {
        // Arrange
        val message = """
            |feat: add multiline support
            |
            |This commit adds support for parsing commit messages
            |that span multiple lines.
        """.trimMargin()

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNotNull().all {
            prop(ConventionalCommit::type).isEqualTo(SectionType.Features)
            prop(ConventionalCommit::description).isEqualTo("add multiline support")
            prop(ConventionalCommit::scope).isNull()
        }
    }

    @Test
    fun `given multiline commit with scope and body when parsing then parses header and ignores body`() {
        // Arrange
        val message = """
            |fix(parser): handle trailing whitespace
            |
            |The parser previously failed when the description had
            |trailing whitespace. This commit trims it.
            |
            |Closes #321
        """.trimMargin()

        // Act
        val result = parser.parse(message, repoUrl)

        // Assert
        assertThat(result).isNotNull().all {
            prop(ConventionalCommit::type).isEqualTo(SectionType.BugFixes)
            prop(ConventionalCommit::scope).isEqualTo("parser")
            prop(ConventionalCommit::description).isEqualTo("handle trailing whitespace")
        }
    }

    @Test
    fun `given multiline commit with header not on first line when parsing then returns null`() {
        // Arrange
        // The conventional commit header is present, but in the wrong order:
        // it appears after a free-form line instead of being the first line.
        val message = """
            |This is a free-form description line.
            |feat: add feature described in the body
        """.trimMargin()

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNull()
    }

    @Test
    fun `given multiline commit with blank first line then header when parsing then returns null`() {
        // Arrange
        // A leading blank line pushes the header off the start of the message.
        val message = "\nfeat: add feature after blank line"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNull()
    }

    // endregion

    // region Breaking change marker

    @Test
    fun `given commit with breaking marker after type when parsing then parses type and excludes marker`() {
        // Arrange
        val message = "feat!: drop support for legacy API"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNotNull().all {
            prop(ConventionalCommit::type).isEqualTo(SectionType.Features)
            prop(ConventionalCommit::scope).isNull()
            prop(ConventionalCommit::description).isEqualTo("drop support for legacy API")
        }
    }

    @Test
    fun `given commit with breaking marker after scope when parsing then parses scope and excludes marker`() {
        // Arrange
        val message = "feat(api)!: drop support for legacy API"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNotNull().all {
            prop(ConventionalCommit::type).isEqualTo(SectionType.Features)
            prop(ConventionalCommit::scope).isEqualTo("api")
            prop(ConventionalCommit::description).isEqualTo("drop support for legacy API")
        }
    }

    // endregion

    // region Casing

    @Test
    fun `given commit with uppercase type when parsing then normalizes type and preserves scope casing`() {
        // Arrange
        val message = "Fix(UI): resolve crash"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNotNull().all {
            prop(ConventionalCommit::type).isEqualTo(SectionType.BugFixes)
            prop(ConventionalCommit::scope).isEqualTo("UI")
            prop(ConventionalCommit::description).isEqualTo("resolve crash")
        }
    }

    // endregion

    // region Description edge-cases

    @Test
    fun `given commit with colon in description when parsing then keeps full description`() {
        // Arrange
        val message = "feat: add support for key: value pairs"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNotNull().all {
            prop(ConventionalCommit::type).isEqualTo(SectionType.Features)
            prop(ConventionalCommit::description).isEqualTo("add support for key: value pairs")
        }
    }

    @Test
    fun `given commit with no space after colon when parsing then parses description`() {
        // Arrange
        val message = "feat:add feature"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNotNull().all {
            prop(ConventionalCommit::type).isEqualTo(SectionType.Features)
            prop(ConventionalCommit::description).isEqualTo("add feature")
        }
    }

    @Test
    fun `given commit with only whitespace description when parsing then returns empty description`() {
        // Arrange
        val message = "feat: "

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNotNull().all {
            prop(ConventionalCommit::type).isEqualTo(SectionType.Features)
            prop(ConventionalCommit::description).isEqualTo("")
        }
    }

    @Test
    fun `given commit with type and colon but no description when parsing then returns null`() {
        // Arrange
        val message = "feat:"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNull()
    }

    // endregion

    // region Malformed headers

    @Test
    fun `given commit with type but no colon when parsing then returns null`() {
        // Arrange
        val message = "feat add feature"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNull()
    }

    @Test
    fun `given commit with empty scope parentheses when parsing then returns null`() {
        // Arrange
        val message = "feat(): add feature"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNull()
    }

    @Test
    fun `given commit with leading whitespace when parsing then returns null`() {
        // Arrange
        val message = "  feat: add feature"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNull()
    }

    @Test
    fun `given well-formed commit with unknown type when parsing then returns null`() {
        // Arrange
        val message = "wip: work in progress"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNull()
    }

    @Test
    fun `given commit with leading words before a valid header when parsing then returns null`() {
        // Arrange
        // A valid-looking "perf:" header appears, but free-form words precede it on
        // the same line. The pattern is anchored to the start of the message, so the
        // header is never reached.
        val message = "wrong format perf: speed up parsing"

        // Act
        val result = parser.parse(message)

        // Assert
        assertThat(result).isNull()
    }

    // endregion

    // region PR linking edge-cases

    @Test
    fun `given commit with PR number without parentheses when parsing then does not link PR`() {
        // Arrange
        val message = "feat: add feature referencing #123"

        // Act
        val result = parser.parse(message, repoUrl)

        // Assert
        assertThat(result).isNotNull().prop(ConventionalCommit::description).isEqualTo("add feature referencing #123")
    }

    @Test
    fun `given repoUrl with trailing slash when parsing then links PR without double slash`() {
        // Arrange
        val message = "feat: add feature (#123)"

        // Act
        val result = parser.parse(message, repoUrl = "$repoUrl/")

        // Assert
        assertThat(result).isNotNull().prop(ConventionalCommit::description).isEqualTo(
            "add feature ([#123](https://github.com/thunderbird/thunderbird-mobile-components/pull/123))",
        )
    }

    // endregion

    // region Type mapping coverage

    @Test
    fun `given docs commit when parsing then maps to Documentation`() {
        assertThat(parser.parse("docs: update readme"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Documentation)
    }

    @Test
    fun `given style commit when parsing then maps to Styles`() {
        assertThat(parser.parse("style: reformat code"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Styles)
    }

    @Test
    fun `given refactor commit when parsing then maps to Refactoring`() {
        assertThat(parser.parse("refactor: extract helper"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Refactoring)
    }

    @Test
    fun `given perf commit when parsing then maps to Refactoring`() {
        assertThat(parser.parse("perf: speed up parsing"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Refactoring)
    }

    @Test
    fun `given test commit when parsing then maps to Tests`() {
        assertThat(parser.parse("test: add parser tests"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Tests)
    }

    @Test
    fun `given revert commit when parsing then maps to Reverts`() {
        assertThat(parser.parse("revert: revert previous change"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Reverts)
    }

    @Test
    fun `given build commit when parsing then maps to Chores`() {
        assertThat(parser.parse("build: bump gradle"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Chores)
    }

    @Test
    fun `given ci commit when parsing then maps to Chores`() {
        assertThat(parser.parse("ci: fix pipeline"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Chores)
    }

    @Test
    fun `given deps commit when parsing then maps to Chores`() {
        assertThat(parser.parse("deps: bump kotlin"))
            .isNotNull().prop(ConventionalCommit::type).isEqualTo(SectionType.Chores)
    }

    // endregion
```

File: docs/commit-guide.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+- ✅ Consider the commit as a documentation tool.
+- ❌ Avoid jargon, acronyms, or vague messages like `update stuff`.
+
+## Summary
+
+- Use [Conventional Commits](#-conventional-commits) for consistency.
+- Keep commit messages short, structured, and focused.
+- Make each commit purposeful and self-contained.
+- Write commits that make collaboration and future development easier for everyone—including you.
+
```
Comment (rafaeltonholo) [inline docs/commit-guide.md]: Shouldn't this be at the top of the document?

File: docs/commit-guide.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
+# Git Commit Guide
+
+Use [Conventional Commits](https://www.conventionalcommits.org/) to write consistent and meaningful commit messages.
+This makes your work easier to review, track, and maintain for everyone involved in the project.
+
+## ✍️ Commit Message Format
```
Comment (rafaeltonholo) [inline docs/commit-guide.md:13]: Missing the mentioning of the `!` for breaking changes

File: docs/release-guide.md [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+## Release Workflow
+
+1. Run the changelog update task for the component being released.
+2. Review the generated `Unreleased` changelog entries.
+3. Apply the release version changes.
+4. Finalize the changelog for the release version.
+5. Open a release pull request with the changelog and version changes.
+6. Merge the release pull request.
+7. Create the component release tag from the merged release commit.
+8. Publish the release.
```
Comment (rafaeltonholo) [inline docs/release-guide.md]:
```suggestion
1. Create the release branch following the regex pattern: `^release\/([a-zA-Z-]+-\d+\.\d+\.\d+)+$` from an up-to-date `main` branch, where:.
    1. The branch name starts with `release/`
    2. The following should be: `<component>-<version>`
    3. In case of multiple components, separete them with `-`
2. Run the changelog update task for the component being released.
3. Review the generated `Unreleased` changelog entries.
4. Apply the release version changes.
5. Finalize the changelog for the release version.
6. Open a release pull request with the changelog and version changes.
7. Merge the release pull request.
8. Create the component release tag from the merged release commit.
9. Publish the release.
```

**Review by rafaeltonholo — COMMENTED** [review]
Thanks for these changes! I've added a few comments, please take a look

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParser.kt (hunk 8 above)
Comment (wmontwe) [reply build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParser.kt:54]: I think we don't need `uplift` for the components. There will be just `main` and version tags.

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/ChangelogPlugin.kt (hunk 2 above)
Comment (wmontwe) [reply build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/ChangelogPlugin.kt]: I think this should be `debug`, as it's only helpful when developing the plugin.

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/internal/Version.kt (hunk 4 above)
Comment (wmontwe) [reply build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/versioning/internal/Version.kt:11]: I switched to git tags to determine if a version is a snapshot.

File: build-plugin/plugin/src/test/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParserTest.kt (hunk 9 above)
Comment (wmontwe) [reply build-plugin/plugin/src/test/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParserTest.kt:126]: hmm, this could be a separate patch. I also wonder if it's better to validate the commits on PRs to avoid extensive tests here.

File: docs/commit-guide.md (hunk 10 above)
Comment (wmontwe) [reply docs/commit-guide.md]: This is a 1:1 copy of https://github.com/thunderbird/thunderbird-android/blob/main/docs/contributing/git-commit-guide.md

But could be at the top.

File: docs/commit-guide.md (hunk 11 above)
Comment (wmontwe) [reply docs/commit-guide.md:13]: So far we don't use breaking, I wonder if it should be on commit or pr level.

File: build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/ChangelogPlugin.kt (hunk 2 above)
Comment (rafaeltonholo) [reply build-plugin/plugin/src/main/kotlin/net/thunderbird/gradle/plugin/changelog/ChangelogPlugin.kt]: So it is fine leaving it as info. Gradle "debug" is too noisy

File: build-plugin/plugin/src/test/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParserTest.kt (hunk 9 above)
Comment (rafaeltonholo) [reply build-plugin/plugin/src/test/kotlin/net/thunderbird/gradle/plugin/changelog/internal/git/GitConventionalCommitParserTest.kt:126]: There may have a GitHub Action that do this. We could experiment a few one

**Review by rafaeltonholo — APPROVED** [review]
LGTM
